File: framework/ui/src/io/intino/alexandria/ui/displays/components/TaskSwitches.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.events.operation.ToggleEvent;
import io.intino.alexandria.ui.displays.events.operation.ToggleEvent.State;
import io.intino.alexandria.ui.displays.events.operation.ToggleListener;
import io.intino.alexandria.ui.displays.notifiers.TaskSwitchesNotifier;

public class TaskSwitches<DN extends TaskSwitchesNotifier, B extends Box> extends AbstractTaskSwitches<DN, B> {
    private State state;
    private ToggleListener toggleListener;

    public TaskSwitches(B box) {
        super(box);
    }

    public TaskSwitches onToggle(ToggleListener listener) {
        this.toggleListener = listener;
        return this;
    }

    public State state() {
        return this.state;
    }

    public TaskSwitches state(State state) {
        _state(state);
        notifyToggle();
        return this;
    }

    public void toggle() {
        state = state == State.On ? State.Off : State.On;
        notifyToggle();
    }

    protected TaskSwitches _state(State state) {
        this.state = state;
        return this;
    }

    private void notifyToggle() {
        notifier.refreshState(state.name());
        if (this.toggleListener == null) return;
        this.toggleListener.accept(new ToggleEvent(this, state));
    }

}

File: framework/ui-elements/src/displays/components/TaskSwitches.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import { Switch, FormControlLabel } from '@material-ui/core';
import AbstractTaskSwitches from "../../../gen/displays/components/AbstractTaskSwitches";
import TaskSwitchesNotifier from "../../../gen/displays/notifiers/TaskSwitchesNotifier";
import TaskSwitchesRequester from "../../../gen/displays/requesters/TaskSwitchesRequester";
import DisplayFactory from 'alexandria-ui-elements/src/displays/DisplayFactory';
import { withSnackbar } from 'notistack';

const styles = theme => ({});

class TaskSwitches extends AbstractTaskSwitches {

	constructor(props) {
		super(props);
		this.notifier = new TaskSwitchesNotifier(this);
		this.requester = new TaskSwitchesRequester(this);
		this.state.checked = this.props.state === "On";
	};

	renderTrigger = () => {
		const {classes} = this.props;
		return (
			<FormControlLabel
				control={<Switch checked={this.state.checked} onChange={this.handleChange.bind(this)}/>}
				label={this._title()}
			/>
		);
	};

	refreshState = (value) => {
		this.setState({ checked: value === "On"});
	};

	handleChange = () => {
		this.requester.toggle();
	};
}

export default withStyles(styles, { withTheme: true })(withSnackbar(TaskSwitches));
DisplayFactory.register("TaskSwitches", withStyles(styles, { withTheme: true })(withSnackbar(TaskSwitches)));